### ISSUE TITLE
An XOR Example

### ISSUE THREAD

**Issue by plaurent** [hpe-cct/cct-nn#26]
Here I'm trying to solve XOR with an MLP.  I'm getting what looks to be a fairly simple error from the CrossEntropySoftMax() function, but it's not clear how to resolve it.

The error is:

`requirement failed: inference field type (VectorField( )( 4 )) must match labels field type VectorField( 1 1 )( 4 )`

Here is the source code (I'm using the handy CSV reader from Issue https://github.com/hpe-cct/cct-nn/issues/4):

```
package tutorial.toolkit.neuralnetwork

import java.io.File

import cogdebugger.CogDebuggerApp
import libcog._
import toolkit.neuralnetwork.function._
import toolkit.neuralnetwork.policy.StandardLearningRule
import toolkit.neuralnetwork.source.{CSVDataSource}
import toolkit.neuralnetwork.util.{CorrectCount, NormalizedLowPass}
import toolkit.neuralnetwork.Implicits._
import toolkit.neuralnetwork.layer.FullyConnectedLayer


/*

inputs.csv
1,0
0,1
1,1
0,0

outputs.csv
1
1
0
0

*/

object MLP extends CogDebuggerApp(new ComputeGraph {
  val learningRule = StandardLearningRule(0.01f, 0.9f, 0.0005f)

  val batchSize = 4;
  val inputData = CSVDataSource("resources/inputs.csv", batchSize, header=false, areNums=true, elemSize=Shape(2,1));
  val outputData = CSVDataSource("resources/outputs.csv", batchSize, header=false, areNums=true, elemSize=Shape(1,1));

  val fc1 = FullyConnectedLayer(inputData, 3, learningRule)
  val r1 = ReLU(fc1)
  val fc2 = FullyConnectedLayer(r1, 1, learningRule)

  val loss = CrossEntropySoftmax(fc2, outputData)
  val normLoss = loss / batchSize
  val correct = CorrectCount(fc2.forward, outputData.forward, batchSize, 0.01f) / batchSize
  val avgCorrect = NormalizedLowPass(correct, 0.001f)
  normLoss.activateSGD()

  probe(fc1.weights.forward)
  fc1.weights.backward.foreach(probe(_, "fc1 update"))

  probe(fc2.weights.forward)
  fc2.weights.backward.foreach(probe(_, "fc2 update"))

  probe(normLoss.forward)
  probe(avgCorrect)
})
```

I get the following error:

```
[info] Running tutorial.toolkit.neuralnetwork.MLP
[error] Exception in thread "main" java.lang.ExceptionInInitializerError
[error]         at tutorial.toolkit.neuralnetwork.MLP.main(MLP.scala)
[error] Caused by: java.lang.IllegalArgumentException: requirement failed: inference field type (VectorField( )( 4 )) must match labels field type VectorField( 1 1 )( 4 )
[error]         at scala.Predef$.require(Predef.scala:219)
[error]         at toolkit.neuralnetwork.function.CrossEntropySoftmax.<init>(CrossEntropySoftmax.scala:28)
[error]         at tutorial.toolkit.neuralnetwork.MLP$$anon$1.<init>(MLP.scala:44)
[error]         at tutorial.toolkit.neuralnetwork.MLP$.<init>(MLP.scala:33)
[error]         at tutorial.toolkit.neuralnetwork.MLP$.<clinit>(MLP.scala)
[error]         ... 1 more
java.lang.RuntimeException: Nonzero exit code returned from runner: 1
        at scala.sys.package$.error(package.scala:27)
[trace] Stack trace suppressed: run last compile:run for the full output.
[error] (compile:run) Nonzero exit code returned from runner: 1
[error] Total time: 2 s, completed Sep 8, 2016 5:41:15 PM
```

**Comment by ghost** [comment]
Just a couple of things. First the final layer of your nn needs to have an outputLen of the number of output classes, in this case it should be 2 not 1. Then second, for your outputData, CCT wants label fields to be Shape() and yours is Shape(1 1) since your using the csv data use the CSVLabelSource instead(Those are on my account not CCT)

**Comment by plaurent** [comment]
@mjfontaine Thanks for the reply.  I've updated my code and it compiles and runs now.  But do you know why there are 8 components on the output rather than 4?

<img width="1074" alt="screen shot 2016-09-09 at 2 25 39 pm" src="https://cloud.githubusercontent.com/assets/1526749/18403387/518ddb48-7699-11e6-9381-12c8589320e7.png">

```
package tutorial.toolkit.neuralnetwork

import java.io.File

import cogdebugger.CogDebuggerApp
import libcog._
import toolkit.neuralnetwork.function._
import toolkit.neuralnetwork.policy.StandardLearningRule
import toolkit.neuralnetwork.source.{CSVDataSource, CSVLabelSource}
import toolkit.neuralnetwork.util.{CorrectCount, NormalizedLowPass}
import toolkit.neuralnetwork.Implicits._
import toolkit.neuralnetwork.layer.FullyConnectedLayer


/*

inputs.csv
1,0
0,1
1,1
0,0

outputs.csv
1
1
0
0

*/

object MLP extends CogDebuggerApp(new ComputeGraph {
  val learningRule = StandardLearningRule(learningRate=0.01f, momentum=0.0f, weightDecay=0.0f)

  val batchSize = 4;
  val inputData = CSVDataSource("resources/inputs.csv", batchSize, header=false, areNums=true, elemSize=Shape(2,1));
  val outputData = CSVLabelSource("resources/outputs.csv", batchSize, header=false, classes=2);

  val layer1 = FullyConnectedLayer(inputData, 3, learningRule)
  val layer1Output = Tanh(layer1)
  val layer2 = FullyConnectedLayer(layer1Output, 2, learningRule)
  val layer2Output = Tanh(layer2)

  val loss = CrossEntropySoftmax(layer2Output, outputData)
  val normLoss = loss / batchSize
  val correct = CorrectCount(layer2Output.forward, outputData.forward, batchSize, 0.01f) / batchSize
  val avgCorrect = NormalizedLowPass(correct, 0.001f)
  normLoss.activateSGD()

  probeAll
})
```

**Comment by plaurent** [comment]
Oh-- I see why there are 8, because of the 1-hot encoding for the labels.  

However, it's not quite learning. Do you know why all the squares in outputData.forward are black?

**Comment by ghost** [comment]
I think once you get down to a 0-dimension field the graphic isn't really helpful to look at. I usually just read the index below it with Floating Max selected

**Comment by bchandle** [comment]
All the squares for outputData.forward are black because all the values are constants. For vectors, the GUI sets the visualization range per plane. That behavior is tuned more towards things like images or intermediate convolutional layer activations in a neural network. If you're trying to visualize a constant vector in a 0-D field, there's only a single value per visualization tile. You'll notice components 1, 3, 4, and 6 have a range from 1.000 to 1.000 while the others have a range from 0.000 to 0.000.

One general tip: for a network this small, the "probeAll" method is a useful tool. That will automatically probe all named fields without requiring you to enumerate them yourself.

The only changes I had to make to get your code to learn properly were adding a bias layer after fc1 and switching to a hyperbolic tangent nonlinearity. ReLU is a fairly weak nonlinear function. It doesn't tend to work very well on tiny fully-connected networks. Here's the modified network:

``` scala
val fc1 = FullyConnectedLayer(inputData, 3, learningRule)
val b1 = BiasLayer(fc1, learningRule)
val r1 = Tanh(b1)
val fc2 = FullyConnectedLayer(r1, 2, learningRule)
```

**Comment by plaurent** [comment]
Fantastic, thank you @mjfontaine and @bchandle -- this works perfectly.  Onward!

For reference, here is the complete XOR example

```
object MLP extends CogDebuggerApp(new ComputeGraph {
  val learningRule = StandardLearningRule(learningRate=0.1f, momentum=0.0f, weightDecay=0.0f)

  val batchSize = 4;
  val inputData = CSVDataSource("resources/inputs.csv", batchSize, header=false, areNums=true, elemSize=Shape(2,1));
  val outputData = CSVLabelSource("resources/outputs.csv", batchSize, header=false, classes=2);

  val fc1 = FullyConnectedLayer(inputData, 3, learningRule)
  val b1 = BiasLayer(fc1, learningRule)
  val r1 = Tanh(b1)

  val fc2 = FullyConnectedLayer(r1, 2, learningRule)

  val loss = CrossEntropySoftmax(fc2, outputData)
  val normLoss = loss / batchSize
  val correct = CorrectCount(fc2.forward, outputData.forward, batchSize, 0.01f) / batchSize
  val avgCorrect = NormalizedLowPass(correct, 0.001f)
  normLoss.activateSGD()

  probeAll
})
```
